### DISCUSSION TITLE
Add Custom TMC setup for Australia and New Zealand

### DISCUSSION THREAD

**Discussion by catatonicChimp** [Mr-MIBonk/M.I.B._More-Incredible-Bash#69] [Ideas]
https://github.com/catatonicChimp/mib2-tmc-ANZ/tree/main/MIB2HIGH(MHI2)
I've created a patch to enable HereTraffic TMC-RDS based on a patch made some people on Drive.ru. It is a set of custom paytmc.json and paytmc_aus.json files with custom entries for HereTraffic in Australia and NZ and an update to /mnt/app/navigation/resources/app/vw/*/version-cfg.txt so the nav unit points to the custom version rather then what was installed via the map update. 

It would be great to add a menu option in M.I.B. to install them. 
the install script probably needs some clean up and currently its only tested with my VW Golf, but have tested it works in Both NZ and NSW (Australia) 
Cheers

**Comment by elysion0619** [comment]
I also look forward to integrating TMC cracking into the new version of MIB. After audi connect expires, VC can no longer display traffic conditions. Currently I see someone on the Russian forum has achieved it.

**Reply by harman-f** [reply]
what you region?

Traffic data from Connect will always be better than TMC

**Reply by elysion0619** [reply]
I am in China, currently the map in the VW car machine (275C, non-MIB2, no SIM card) can display TMC information after cracking, I hope AUDI can also achieve it, if you need car machine files or testing I can help.

**Comment by harman-f** [comment]
If you can bring the input and help with testing. 
sure 😊

**Comment by catatonicChimp** [comment]
When I find some time in the next few weeks I write up a proper script and a menu for it (now that I sunk some time into understanding the menus last week) AUS/NZ is easy as I have that data, and I could probably find the Russian info as that is where I started with this and have the original configs somewhere.

It does rely on two things, one that the encryption keys exist in the encryption tables already (via maps or base firmware) it's not always the case and in AUS/NZ the keys were removed from the discover media maps in late 2010s, though still works fine with MIB Discover pro units etc, and then a little bit of info on which stations actually broadcast the tmc data (not hard to find with a sdr radio or similar) Station Id, country id etc

The last thing is just trying the "7" different encryption index values until you find what works.

**Reply by harman-f** [reply]
no need to get in that much detail, I can handle the scripts or GEM menu.
I just need the values ;)

I already started this:
```
BIOSCtrl
	value	int per 28442848 520
	label	"Set PayTMC"
	entry	"none" 0
	entry	"DE/EU" 32771 
	entry	"China" 32783
	poll	1000
```
This is for display and configuration in GEM

**Reply by Jguog** [reply]
Hi really appreciate the work gone into this. I know it's been a while. I'm in Aus, looking to apply this. Has there been any progress on integrating it? Or any instructions on how to run it manually. Is this just adding HERETraffic as a source or will it also enable online traffic data too?

**Reply by catatonicChimp** [reply]
Last I checked it was already integrated

**Comment by djskual** [comment]
guys, thx a lot for your hard work.
Hope you will find a way to make TMC work everywhere and integrarte this in M.I.B.
If i can help for France, tell me the way

**Reply by catatonicChimp** [reply]
Hey @djskual 

I was pretty sure france already has configs? have you coded/adapted your radio to point to the "fee-based traffic" for france?
Value: 20 or 32788
#             20
		{
			"SetId" : 20,
			"Name"  : "PayTMC France",
			"ProviderSet" : [
				{
					"Name"  : "free",
					"Value" : "yes"
				},
				{
					"Name"  : "mediamobile",
					"Value" : "yes"
				}
			]
		},
		
#				32768 + 20
		{
			"SetId" : 32788,
			"Name"  : "online traffic + PayTMC France",
			"ProviderSet" : [
				{
					"Name"  : "free",
					"Value" : "yes"
				},
				{
					"Name"  : "mediamobile",
					"Value" : "yes"
				},
				{
					"Name"  : "onlinetraffic",
					"Value" : "yes"
				}
			]
		},
		
#		Mediamobile France
		{
			"Name" : "mediamobile",
			"Type" : "unencrypted",
			"SDKItem" :[
				{
					"Name"  : "WorldRegions",
					"Nr"    : 1224736801,
					"Value" : 1
				},
				{
					"Name"  : "TableNumbers",
					"Nr"    : 1224736802,
					"Value" : 30
				},
				{
					"Name"  : "CountryCodes",
					"Nr"    : 1224736803,
					"Value" : 15
				},
				{
					"Name"  : "unencrypted ExtendedCountryCodes",
					"Nr"    : 1224736830,
					"Value" : 225
				}
			]
		},

#		ViaMichelin
		{
			"Name" : "viamichelin",
			"Type" : "unencrypted",
			"SDKItem" :[
				{
					"Name"  : "WorldRegions",
					"Nr"    : 1224736801,
					"Value" : 1
				},
				{
					"Name"  : "TableNumbers",
					"Nr"    : 1224736802,
					"Value" : 29
				},
				{
					"Name"  : "CountryCodes",
					"Nr"    : 1224736803,
					"Value" : 15
				},
				{
					"Name"  : "unencrypted ExtendedCountryCodes",
					"Nr"    : 1224736830,
					"Value" : 225
				}
			]
		},

**Reply by djskual** [reply]
OK, i will try the 32788, i seen my radio is coded with 32771.

**Reply by djskual** [reply]
OK, i tested, 32788 is working for me in france. original 32771 doesn't work. Thanks a lot.

**Reply by harman-f** [reply]
Strange, thougth that 32771 should work in most of europe

```
BIOSCtrl
	value	int per 28442848 520
	label	"Set PayTMC"
	entry	"none" 0
	entry	"DE/EU" 32771
	entry	"France" 32788
	entry	"China" 32783
	poll	1000
```

Option will be added to next release

**Reply by catatonicChimp** [reply]
it looks like lots a lot of europe has fragmented over time, as part of the paytmc.json file edits, we could add an Uber entry with every single tmc provider. I don't think there would be any issues with it, though those close to borders might get excess notifications... hard to say without been in europe to test.. 

regarding france, I am surprised that viaMichelin isn't in the 32771 id, according to wikipedia they do exist in france... 



Here is a list of all the IDs if you adding them to a menu. 
```

"SetId" : 0,
			"Name"  : "free",
"SetId" : 1,
			"Name"  : "Audi TMCpro",
"SetId" : 2,
			"Name"  : "AUDI w/o TMCPro",
"SetId" : 3,
			"Name"  : "VW TMCPro",
"SetId" : 4,
			"Name"  : "VW w/o TMCPro",
"SetId" : 5,
			"Name"  : "Bentley TMCPro",
"SetId" : 6,
			"Name"  : "Bentley w/o TMCPro",
"SetId" : 7,
			"Name"  : "Bentley w/o TMCPro",
"SetId" : 8,
			"Name"  : "AUDI TMCPro + infoBlu",
"SetId" : 9,
			"Name"  : "AUDI w/o TMCPro + infoBlu",
"SetId" : 10,
			"Name"  : "VW TMCPro + infoBlu",
"SetId" : 11,
			"Name"  : "VW w/o TMCPro + infoBlu",
"SetId" : 12,
			"Name"  : "Bentley TMCPro + infoBlu",
"SetId" : 13,
			"Name"  : "Bentley w/o TMCPro + infoBlu",
"SetId" : 14,
			"Name"  : "TMCPro + itis + infoBlu + viaMichelin",
"SetId" : 15,
			"Name"  : "TMCPro + trafficmaster + infoBlu + viaMichelin",
"SetId" : 16,
			"Name"  : "TMCPro + trafficmaster + infoBlu + viaMichelin + freeTM russia",
"SetId" : 17,
			"Name"  : "TMCPro + trafficmaster + infoBlu + viaMichelin + TMCpro Russia",
"SetId" : 18,
			"Name"  : "TMCPro + trafficmaster + infoBlu + viaMichelin + freeDAB-TPEG",
"SetId" : 19,
			"Name"  : "PayTMC GER",
"SetId" : 20,
			"Name"  : "PayTMC France",
"SetId" : 21,
			"Name"  : "PayTMC UK",
"SetId" : 22,
			"Name"  : "PayTMC PT",
"SetId" : 23,
			"Name"  : "PayTMC IT",
"SetId" : 24,
			"Name"  : "PayTMC FI",
"SetId" : 25,
			"Name"  : "trafficmaster + infoBlu + viaMichelin + finlanddestia",
"SetId" : 26,
			"Name"  : "Skoda freeTMC+TMCpro+Trafficmaster+ViaMichelin+MediaMobileNordc+TrafficNav",
"SetId" : 27,
			"Name"  : "Skoda freeTMC+TMCpro+Trafficmaster+ViaMichelin+infoBlu+MediaMobileNordc+TrafficNav",
"SetId" : 1023,
			"Name"  : "VW PayTMC all",
"SetId" : 1024,
			"Name"  : "SUNA Traffic (Intelematics)",
"SetId" : 1025,
			"Name"  : "Altech Netstar",
"SetId" : 1536,
			"Name"  : "NAR SDARS Sirius/XM + FreeTMC",
"SetId" : 1537,
			"Name"  : "NAR SDARS Sirius/XM",
"SetId" : 1538,
			"Name"  : "NAR Clearchannel HDRadio",
"SetId" : 32768,
			"Name"  : "online traffic + only free TMC",
"SetId" : 32769,
			"Name"  : "online traffic + AUDI TMCPro",
"SetId" : 32770,
			"Name"  : "online traffic + AUDI w/o TMCPro",
"SetId" : 32771,
			"Name"  : "online traffic + VW TMCPro",
"SetId" : 32772,
			"Name"  : "online traffic + VW w/o TMCPro",
"SetId" : 32773,
			"Name"  : "online traffic + Bentley TMCPro",
"SetId" : 32774,
			"Name"  : "online traffic + Bentley w/o TMCPro",
"SetId" : 32775,
			"Name"  : "online traffic + Bentley w/o TMCPro",
"SetId" : 32776,
			"Name"  : "online traffic + AUDI TMCPro + infoBlu",
"SetId" : 32777,
			"Name"  : "online traffic + AUDI w/o TMCPro + infoBlu",
"SetId" : 32778,
			"Name"  : "online traffic + VW TMCPro + infoBlu",
"SetId" : 32779,
			"Name"  : "online traffic + VW w/o TMCPro + infoBlu",
"SetId" : 32780,
			"Name"  : "online traffic + Bentley TMCPro + infoBlu",
"SetId" : 32781,
			"Name"  : "online traffic + Bentley w/o TMCPro + infoBlu",
"SetId" : 32782,
			"Name"  : "online traffic + Bentley w/o TMCPro + itis + viaMichelin + infoBlu",
"SetId" : 32783,
			"Name"  : "online traffic + Bentley w/o TMCPro + trafficmaster + viaMichelin + infoBlu",
"SetId" : 32784,
			"Name"  : "online traffic + TMCPro + trafficmaster + infoBlu + viaMichelin + freeTM russia",
"SetId" : 32785,
			"Name"  : "online traffic + TMCPro + trafficmaster + infoBlu + viaMichelin + TMCpro Russia",
"SetId" : 32786,
			"Name"  : "online traffic + TMCPro + trafficmaster + infoBlu + viaMichelin + freeDAB-TPEG",
"SetId" : 32787,
			"Name"  : "online traffic + PayTMC GER",
"SetId" : 32788,
			"Name"  : "online traffic + PayTMC France",
"SetId" : 32789,
			"Name"  : "online traffic + PayTMC UK",
"SetId" : 32790,
			"Name"  : "online traffic + PayTMC PT",
"SetId" : 32791,
			"Name"  : "online traffic + PayTMC IT",
"SetId" : 32792,
			"Name"  : "online traffic + PayTMC FI",
"SetId" : 32793,
			"Name"  : "online traffic + trafficmaster + infoBlu + viaMichelin + finlanddestia",
"SetId" : 32794,
			"Name"  : "Skoda freeTMC+TMCpro+Trafficmaster+ViaMichelin+MediaMobileNordc+TrafficNav",
"SetId" : 32795,
			"Name"  : "Skoda freeTMC+TMCpro+Trafficmaster+ViaMichelin+infoBlu+MediaMobileNordc+TrafficNav",
"SetId" : 33791,
			"Name"  : "online traffic + VW PayTMC all",
"SetId" : 33792,
			"Name"  : "online traffic + SUNA Traffic (Intelematics)",
"SetId" : 33793,
			"Name"  : "online traffic + Altech Netstar",
"SetId" : 34304,
			"Name"  : "online traffic + NAR SDARS Sirius/XM + FreeTMC",
"SetId" : 34305,
			"Name"  : "online traffic + NAR SDARS Sirius/XM",
"SetId" : 34306,
			"Name"  : "online traffic + NAR Clearchannel HDRadio",
"SetId" : 65533,
			"Name"  : "Online traffic only",
"SetId" : 65534,
			"Name"  : "No TMC + Mecca indicator"
"SetId" : 65535,
			"Name"  : "No TMC"
```

**Reply by catatonicChimp** [reply]
```
{
	"SetId" : 8888, #can be any number not already used
	"Name"  : "Uber List of TMC channels",
	"ProviderSet" : [
        {
          "Name"  : "free",
          "Value" : "yes"
        },
        {
          "Name"  : "tmcpro",
          "Value" : "exclusive"
        },
        {
          "Name"  : "trafficmaster",
          "Value" : "exclusive"
        },
        {
          "Name"  : "mediamobile",
          "Value" : "yes"
        },
        {
          "Name"  : "viamichelin",
          "Value" : "yes"
        },
        {
          "Name"  : "infoblu",
          "Value" : "yes"
        },
        {
          "Name"  : "itis",
          "Value" : "yes"
        },
        {
          "Name"  : "finlanddestia",
          "Value" : "yes"
        },
        {
          "Name"  : "Hungary (TraficNav) unencrypted",
          "Value" : "yes"
        },
        {
          "Name"  : "Hungary (TraficNav) encrypted",
          "Value" : "yes"
        },
        {
          "Name"  : "russiafreetmc",
          "Value" : "yes"
        },
        {
          "Name"  : "russianavteq",
          "Value" : "yes"
        },
        {
          "Name"  : "freeDAB-TPEG",
          "Value" : "yes"
        },
        {
          "Name"  : "Portugal BeMobile",
          "Value" : "yes"
        },
        {
          "Name"  : "Captial Territory (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "New South Wales (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Victoria (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Queensland,Brisbane,Gold Coast (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "South Australia (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Western Australia (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Tasmania (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Northern Territory (AUS)",
          "Value" : "yes"
        },
        {
          "Name"  : "Aukland, Wellington, Christchurch (NZ)",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic ACT",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic NSW",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic VIC",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic QLD",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic SA",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic WA",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic TAS",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic NT",
          "Value" : "yes"
        },
        {
          "Name"  : "HereTraffic NZ",		
          "Value" : "yes"
        },
        {
          "Name"  : "Southafrica Altech Netstar",
          "Value" : "yes"
        },
        {
          "Name"  : "NAR Sirius",
          "Value" : "yes"
        },
        {
          "Name"  : "NAR Clearchannel",
          "Value" : "yes"
        },
        {
          "Name"  : "onlinetraffic",
          "Value" : "yes"
        },
    ]
},
```

**Reply by djskual** [reply]
I found this:
[https://github.com/jilleb/mib2-toolbox/wiki/TMC-channels-research](https://github.com/jilleb/mib2-toolbox/wiki/TMC-channels-research)

**Reply by harman-f** [reply]
Yes, not the 1st round on this topic ;)

<html><body>
<h2>EU Firmware</h2>
<h3>TMC channels without online traffic</h3>


Value | Online Traffic | freeDAB-TPEG | FreeTMC | FreeTMC (RU) | TMCpro (DE) | TMCpro Trafficmaster (UK) | TMCpro MediaMobile (FR) | TMCpro viaMichelin (EU) | TMCpro infoBlue (IT) | TMCpro itis (CN) | TMCpro Navteq (RU) | TMCpro BeMobile (PT) | TMCpro TraficNav (HU) | TMCpro (FI) | TMCpro Intelematics (AUZ/NZ) | TMCpro Altech Netstar (ZAF) | Sirius/XM (US) | HDRadio (US)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 |   |   | X |   |   |   |   |   |   |   |   |   |   |   |   |   |   |  
1 |   |   | X |   | X | X | X |   |   |   |   |   |   |   |   |   |   |  
2 |   |   | X |   |   | X | X |   |   |   |   |   |   |   |   |   |   |  
3 |   |   | X |   | X* | X |   | X |   |   |   |   |   |   |   |   |   |  
4 |   |   | X |   |   | X |   | X |   |   |   |   |   |   |   |   |   |  
5 |   |   | X |   | X | X | X |   |   |   |   |   |   |   |   |   |   |  
6 |   |   | X |   |   | X | X |   |   |   |   |   |   |   |   |   |   |  
7 |   |   | X |   |   | X* | X |   |   |   |   |   |   |   |   |   |   |  
8 |   |   | X |   | X | X | X |   | X |   |   |   |   |   |   |   |   |  
9 |   |   | X |   |   | X | X |   | X |   |   |   |   |   |   |   |   |  
10 |   |   | X |   | X* | X |   | X | X |   |   |   |   |   |   |   |   |  
11 |   |   | X |   |   | X |   | X | X |   |   |   |   |   |   |   |   |  
12 |   |   | X |   | X | X | X |   | X |   |   |   |   |   |   |   |   |  
13 |   |   | X |   |   | X | X |   | X |   |   |   |   |   |   |   |   |  
14 |   |   | X |   | X* |   |   | X | X | X |   |   |   |   |   |   |   |  
15 |   |   | X |   | X* | X |   | X | X |   |   |   | X | X |   |   |   |  
16 |   |   | X | X | X* | X |   | X | X |   |   |   |   |   |   |   |   |  
17 |   |   | X |   | X* | X |   | X | X |   | X |   |   |   |   |   |   |  
18 |   | X | X |   | X* | X |   | X | X |   |   |   |   |   |   |   |   |  
19 |   |   | X |   | X* |   |   |   |   |   |   |   |   |   |   |   |   |  
20 |   |   | X |   |   |   | X |   |   |   |   |   |   |   |   |   |   |  
21 |   |   | X |   |   | X |   |   |   |   |   |   |   |   |   |   |   |  
22 |   |   | X |   |   |   |   |   |   |   |   | X |   |   |   |   |   |  
23 |   |   | X |   |   |   |   |   | X |   |   |   |   |   |   |   |   |  
24 |   |   | X |   |   |   |   |   |   |   |   |   |   | X |   |   |   |  
25 |   |   | X |   |   | X |   | X | X |   |   |   | X | X |   |   |   |  
26 |   |   | X |   | X* | X |   | X |   |   |   |   | X | X |   |   |   |  
27 |   |   | X |   | X* | X |   | X | X |   |   |   | X | X |   |   |   |  
1023 |   | X | X |   | X* | X |   | X |   |   |   | X |   | X |   |   |   |  
1024 |   |   | X |   |   |   |   |   |   |   |   |   |   |   | X |   |   |  
1025 |   |   | X |   |   |   |   |   |   |   |   |   |   |   |   | X |   |  
1536 |   |   | X |   |   |   |   |   |   |   |   |   |   |   |   |   | X |  
1537 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | X |  
1538 |   |   | X |   |   |   |   |   |   |   |   |   |   |   |   |   |   | X


<p>* Excusive tag is set. Don't really sure what it does depending on just active variable.</p>
</body>
</html>

**Reply by harman-f** [reply]
<html><body>
<h3>PayTMC channels with online traffic</h3>


Value | Online Traffic | freeDAB-TPEG | FreeTMC | FreeTMC (RU) | TMCpro (DE) | TMCpro Trafficmaster (UK) | TMCpro MediaMobile (FR) | TMCpro viaMichelin (EU) | TMCpro infoBlue (IT) | TMCpro itis (CN) | TMCpro Navteq (RU) | TMCpro BeMobile (PT) | TMCpro TraficNav (HU) | TMCpro (FI) | TMCpro Intelematics (AUZ/NZ)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
32768 | X |   | X |   |   |   |   |   |   |   |   |   |   |   |  
32769 | X |   | X |   | X | X | X |   |   |   |   |   |   |   |  
32770 | X |   | X |   |   | X | X |   |   |   |   |   |   |   |  
32771 | X |   | X |   | X* | X |   | X |   |   |   |   |   |   |  
32772 | X |   | X |   |   | X |   | X |   |   |   |   |   |   |  
32773 | X |   | X |   | X | X | X |   |   |   |   |   |   |   |  
32774 | X |   | X |   |   | X | X |   |   |   |   |   |   |   |  
32775 | X |   | X |   |   | X* |   |   |   |   |   |   |   |   |  
32776 | X |   | X |   | X | X | X |   | X |   |   |   |   |   |  
32777 | X |   | X |   |   | X | X |   | X |   |   |   |   |   |  
32778 | X |   | X |   | X* | X |   | X | X |   |   |   |   |   |  
32779 | X |   | X |   |   | X |   | X | X |   |   |   |   |   |  
32780 | X |   | X |   | X | X | X |   | X |   |   |   |   |   |  
32781 | X |   | X |   |   | X | X |   | X |   |   |   |   |   |  
32782 | X |   | X |   | X |   |   | X | X | X |   |   |   |   |  
32783 | X |   | X |   | X | X |   | X | X |   |   |   | X | X |  
32784 | X |   | X | X | X* | X |   | X | X |   |   |   |   |   |  
32785 | X |   | X |   | X* | X |   | X | X |   | X |   |   |   |  
32786 | X | X | X |   | X* | X |   | X | X |   |   |   |   |   |  
32787 | X |   | X |   | X* |   |   |   |   |   |   |   |   |   |  
32788 | X |   | X |   |   |   | X |   |   |   |   |   |   |   |  
32789 | X |   | X |   |   | X |   |   |   |   |   |   |   |   |  
32790 | X |   | X |   |   |   |   |   |   |   |   | X |   |   |  
32791 | X |   | X |   |   |   |   |   | X |   |   |   |   |   |  
32792 | X |   | X |   |   |   |   |   |   |   |   |   |   | X |  
32793 | X |   | X |   |   | X |   | X | X |   |   |   | X | X |  
32794 | X |   | X |   | X* | X |   | X |   |   |   |   | X | X |  
32795 | X |   | X |   | X* | X |   | X | X |   |   |   | X | X |  
33791 | X | X | X |   | X* | X |   | X |   |   |   | X |   | X |  
33792 | X |   | X |   |   |   |   |   |   |   |   |   |   |   | X
65533 | X |   |   |   |   |   |   |   |   |   |   |   |   |   |  


<p>* Excusive tag is set. Don't really sure what it does depending on just active variable.</p>
</body>
</html>

**Comment by neoalius** [comment]
I am in AUS and changed 1024 to  33792, it enables the Traffic overlay menu in Navigation but there is no traffic information available.

**Reply by catatonicChimp** [reply]
Pretty sure I only setup the new hereTraffic configs on 1024, not 33972 (since online traffic has never been a thing in Australia, even when I've connected my car via wifi it won't do anything as the vin isn't registered against the servers.)

**Reply by catatonicChimp** [reply]
Sorry @neoalius, the current version on there didn't have the new configs in there, I've just put a pull request through with updated configs

**Reply by harman-f** [reply]
Pull request is merged.

For some users online traffic might be of interest.
There are custom server solutions around.

```
BIOSCtrl
	value	int per 28442848 520
	label	"Set TMC"
	entry	"none" 0
	entry	"Germany" 32771
	entry	"France" 32788
	entry	"Australia" 1024
	entry	"China" 32782
	entry	"Russia" 32785
	poll	1000
```

**Reply by BooBoo124** [reply]
Hi thanks for your great work. Can a script be written for TMC in the UK for Audi A4 please.

**Reply by catatonicChimp** [reply]
Short answer is no... just because I don't live in the UK and I don't know the configs required to make it work
But if you have a SDR radio or willing to spend some time in the GEM Radio menu to work out the base settings I can help you find the settings that we can add to a config.

For example the current TMC file has the following for the UK with the id of 21, using trafficmaster as the primary source (other than free)
```
		{
			"SetId" : 21,
			"Name"  : "PayTMC UK",
			"ProviderSet" : [
				{
					"Name"  : "free",
					"Value" : "yes"
				},
				{
					"Name"  : "trafficmaster",
					"Value" : "yes"
				}
			]
		},

```

now weirdly the above config points to the "trafficmaster" config, but there is also a UK-Trafficmaster config which is slightly different with different "Nr" values... at one point I did work out what they meant, but that was a while ago and I have forgotten. 

```
		{
			"Name" : "trafficmaster",
			"Type" : "unencrypted",
			"SDKItem" :[
				{
# TMC_FEATURE_SUPPORTED_UNENCRYPTED_PAY_TMC_WORLDREGIONS
					"Name"  : "WorldRegions",
					"Nr"    : 1224736801,
					"Value" : 1
				},
				{
# TMC_FEATURE_SUPPORTED_UNENCRYPTED_PAY_TMC_TABLENUMBERS
					"Name"  : "TableNumbers",
					"Nr"    : 1224736802,
					"Value" : 7
				},
				{
# TMC_FEATURE_SUPPORTED_UNENCRYPTED_PAY_TMC_COUNTRYCODES
					"Name"  : "CountryCodes",
					"Nr"    : 1224736803,
					"Value" : 12
				},
				{
# TMC_FEATURE_SUPPORTED_UNENCRYPTED_PAY_TMC_EXTENDEDCOUNTRYCODES
					"Name"  : "unencrypted ExtendedCountryCodes",
					"Nr"    : 1224736830,
					"Value" : 225
				}
			]
		},
```


(UK entry which isn't used)
```
		{
			"Name" : "UK-Trafficmaster",
			"Type" : "unencrypted",
			"SDKItem" :[
				{
					"Name"  : "WorldRegions",
					"Nr"    : 1224736822,
					"Value" : 1
				},
				{
					"Name"  : "TableNumbers",
					"Nr"    : 1224736823,
					"Value" : 7
				},
				{
					"Name"  : "CountryCodes",
					"Nr"    : 1224736824,
					"Value" : 12
				},
				{
					"Name"	: "online ExtendedCountryCodes",
					"Nr"	: 1224736832,
					"Value"	: 225
				}
			]
		},
```

I'll use an example from my Australian configs... 
```
		{
			"Name" : "New South Wales (AUZ)",
			"Type" : "encrypted",
			"SDKItem" :[
				{
					"Name"	: "SID",
					"Nr"	: 1224736808,
					"Value"	: 1
				},
				{
					"Name"	: "TableNumbers",
					"Nr"	: 1224736806,
					"Value"	: 54
				},
				{
					"Name"	: "CountryCodes",
					"Nr"	: 1224736807,
					"Value"	: 2
				},
                                {
					"Name"  : "encrypted ExtendedCountryCodes",
					"Nr"    : 1224736831,
					"Value" : 240
                                },
				{
					"Name"  : "WorldRegions",
					"Nr"    : 1224736805,
					"Value" : 3
				},
				{
					"Name"	: "SVKTableNr",
					"Nr"	: 1224736809,
					"Value"	: 4
				}
			]
		},
```

A) SID (Station ID) of the station that broadcast encrypted TMC, usually they have 1 ID per TMC provider, in Australia Suna is on PID1, and HereTraffic is 7, you can find this tuning to different station in the GEM menu and see which are carrying TMC data and what ID's they have associated with it, (again it's been a while since I dug into  it, so you are gonna have to dig on your own here a bit)
B) Table Numbers according to here traffic and the existing configs,, Table number is 7 so we can reuse that value
C) Country code.. I checked this against hereTraffic (E1C07  GBR United Kingdom) so 12 is the correct value (Hex C = 12 in decimal) (more details here on how to decode the country code: https://developer.here.com/documentation/traffic/dev_guide/topics/resource-type-tmc-code.html)
D) ExtendedCountry Code (now not sure if this in encrypted or not) E1 = 225 you can try crib how some of the other entries are in the script to try a couple of different options...
E) WorldRegion.. pretty simple.. "europe" = 1 (obviously UK is no longer in the EU but this doesn't matter)
F) SVKTableNr: the most important part of an encrypted service... the encryption key... its not published anywhere but thankfully you can guess it, there is less then 10 values it can be I want to say its 0-8, but just to be sure I would try 0-10...

You can edit the json TMC file on your SD card and try and reload it over and over (reboot between I think and you have to wait 5mins or so for it to update depending on your reception... 

so I would edit the 
```
		{
			"SetId" : 21,
			"Name"  : "PayTMC UK",
			"ProviderSet" : [
				{
					"Name"  : "free",
					"Value" : "yes"
				},
				{
					"Name"  : "trafficmaster",
					"Value" : "yes"
				},
				{
					"Name"  : "UK-Trafficmaster",
					"Value" : "yes"
				},
				{
					"Name"  : "UK-Traffic", #name it whatever you want just make it match the next config
					"Value" : "yes"
				},
			]
		},
```

```
		{
			"Name" : "UK-Traffic",
			"Type" : "encrypted",
			"SDKItem" :[
				{
					"Name"	: "SID",
					"Nr"	: 1224736808,
					"Value"	: 7
				},
				{
					"Name"	: "TableNumbers",
					"Nr"	: 1224736806,
					"Value"	: 7
				},
				{
					"Name"	: "CountryCodes",
					"Nr"	: 1224736807,
					"Value"	: 12
				},
                                {
					"Name"  : "encrypted ExtendedCountryCodes",
					"Nr"    : 1224736831,
					"Value" : 225
                                },
                                {
                                        #Probably not needed but I don't see any harm in adding both
					"Name"  : "unencrypted ExtendedCountryCodes",
					"Nr"    : 1224736830,
					"Value" : 225
                                },
				{
					"Name"  : "WorldRegions",
					"Nr"    : 1224736805,
					"Value" : 1
				},
				{
					"Name"	: "SVKTableNr",
					"Nr"	: 1224736809,
					"Value"	: 7
					#If NZ, Aus and Russia all have HereTraffic on Table 7, I imagine its likely the same is true for the UK, but you never know
				}
			]
		},
```

**Reply by BooBoo124** [reply]
Thanks for the detailed information, but sadly it's all a bit Greek to me. Thanks anyway

**Reply by BooBoo124** [reply]
Thanks for the detailed information, but sadly it's all a bit Greek to me. Thanks anyway3

**Reply by G60Dub** [reply]
I'd be interestd in trying the manual edits to the JSON files for the UK/GB settings. I do however, have a Delphi MIB2 unit with the GEM enabled.  Is there anything to be aware of prior to trying this?

**Comment by huanpuzzz** [comment]
Hi,there
You done great work. 
Can a script be written TMC for Singapore ? for Audi A3 8v mib2 please.

**Comment by cdnninja** [comment]
Ran across this and makes me want to play with the HDradio option again. I am in NA in a place where Kia as an example gets free traffic via HDradio.  I see that code as an option and have tried that on my b9 without any luck.   Must be missing something beyond 1538.
